### PR TITLE
Add WarningsFilter fixture

### PR DESCRIPTION
**Pull request by stephenfin**
This has enough users around OpenStack to justify adding it to 'fixtures' proper. It's intentionally dumb, since the main purpose of this is to avoid people calling `resetwarnings` in their variant of the fixture, as that clears *all* filters including those we don't control.

---
This change is [Reviewable](https://reviewable.io/reviews/testing-cabal/fixtures/50)

### CODE REVIEW

File: fixtures/tests/_fixtures/test_warnings.py [hunk 1]
```diff
@@ -47,3 +47,28 @@ def test_capture_category(self):
...
+            ],
+        )
+        self.useFixture(fixture)
+        with warnings.catch_warnings(record=True) as w:
+            warnings.warn('deprecated', DeprecationWarning)
+            warnings.warn('user', UserWarning)
+
+        # only the user warning should be present, and it should only have been
+        # raised once
+        self.assertEqual(1, len(w))
```
Comment (cjwatson) [inline fixtures/tests/_fixtures/test_warnings.py:74]: Could we have a quick test here that the original filters have been properly restored?  You said that doing this was the main point of having this fixture here, and I agree that's a good reason, so it seems that it would be worthwhile to explicitly test that behaviour.

**Review by cjwatson — APPROVED** [review]
A `NEWS` file entry would be good here as well, if you could.

File: fixtures/tests/_fixtures/test_warnings.py (hunk 1 above)
Comment (stephenfin) [reply fixtures/tests/_fixtures/test_warnings.py:86]: Done.

**Comment by stephenfin** [comment]
Fixed comments. Turns out we weren't actually running any of the tests. This is now resolved.

File: fixtures/tests/_fixtures/test_warnings.py [hunk 2]
```diff
@@ -47,3 +59,60 @@ def test_capture_category(self):
...
+                },
+            ],
+        )
+
+        # we copy the filter values rather than a reference to the containing
+        # list since that can change
+        old_filters = warnings.filters[:]
+
+        # NOTE: we intentionally do not use 'self.useFixture' since we want to
+        # teardown the fixture manually here before we exit this test method
```
Comment (cjwatson) [inline fixtures/tests/_fixtures/test_warnings.py:107]: Fair enough, but consider `with fixture:` instead so that the fixture is still cleaned up if the test fails inside the section of code where it's active.

File: fixtures/tests/_fixtures/test_warnings.py [hunk 3]
```diff
@@ -18,9 +18,13 @@
 import fixtures
 
 
-class TestWarnings(testtools.TestCase, fixtures.TestWithFixtures):
+class TestWarningsCapture(testtools.TestCase, fixtures.TestWithFixtures):
 
     def test_capture_reuse(self):
+        # DeprecationWarnings are hidden by default in Python 3.2+, enable them
+        # https://docs.python.org/3/library/warnings.html#default-warning-filter
+        warnings.simplefilter("always")
```
Comment (cjwatson) [inline fixtures/tests/_fixtures/test_warnings.py:26]: Does this leak configuration into other tests?  Maybe we should just have `tox` run the whole test suite with `PYTHONWARNINGS=always` or something like that?

**Review by cjwatson — APPROVED** [review]

File: fixtures/tests/_fixtures/test_warnings.py (hunk 3 above)
Comment (stephenfin) [reply fixtures/tests/_fixtures/test_warnings.py:26]: Yeah, I'm not sure. Given it's unit tests as opposed to something someone else will import, it didn't seem that big an issue. If we do this in tox, we're reliant on people running tox for unit tests to pass. Would it be weird to use our own fixtures (i.e. the new environment fixture) in the tests for other fixtures?

File: fixtures/tests/_fixtures/test_warnings.py (hunk 2 above)
Comment (stephenfin) [reply fixtures/tests/_fixtures/test_warnings.py:107]: Didn't know that was a thing. Will respin once I get a resolution on the below.

Comment (stephenfin) [reply fixtures/tests/_fixtures/test_warnings.py:107]: Done

File: fixtures/tests/_fixtures/test_warnings.py (hunk 3 above)
Comment (stephenfin) [reply fixtures/tests/_fixtures/test_warnings.py:27]: I ended up using my own fixture.

File: fixtures/tests/_fixtures/__init__.py [hunk 4]
```diff
@@ -1,34 +1,28 @@
...
 # WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 # license you chose for the specific language governing permissions and
 # limitations under that license.
 
+import os
+
+
 def load_tests(loader, standard_tests, pattern):
     test_modules = [
-        'environ',
```
Comment (jelmer) [inline fixtures/tests/_fixtures/__init__.py:18]: This seems unrelated, and perhaps more controversial. Perhaps leave for a separate PR?

**Review by jelmer — CHANGES_REQUESTED** [review]
The core changes look good to me

**Comment by stephenfin** [comment]
Split some unnecessary changes out into a separate PR. Is this good to go now, @jelmer / @cjwatson ?

File: fixtures/tests/_fixtures/__init__.py (hunk 4 above)
Comment (stephenfin) [reply fixtures/tests/_fixtures/__init__.py]: Now part of https://github.com/testing-cabal/fixtures/pull/75

**Review by jelmer — APPROVED** [review]
